text/2020-01-30-key-spaces.md

```diff
@@ -0,0 +1,181 @@
+# Title
+
+Key Spaces
+
+## Summary
+
+When multiple applications store data in TiKV, we need a clear pattern for avoiding conflicting use of the key space in TiKV.
+TiKV is a monolithic KV store, so each application must choose a key prefix.
+Key Spaces formalizes this process by registering a key space for an application in TiKV itself.
+
+When combined with authentication (not part of this proposal), this can secure an application's key space.
+
+Additionally, key spaces make it possible for TiKV to avoid conflicts during a data restore by changing the key prefix assigned to a key space.
+
+## Motivation
+
+The use case for this is running multiple applications against TiKV (not just a single TiDB).
+Note that we may be motivated to run multiple instances of a single application such as TiDB.
+
+* Avoiding even accidental conflicting use of key space in TiKV is currently ad-hoc and unreliable
+* Securing data in TiKV is difficult. A necessary first step is clearly defining the boundaries of data.
+
+Once key spaces are in place, all applications and TiKV itself should understand what data belongs to whom and how to avoid key space conflict.
```

Comment:
I'd like to propose a client-only alternative. Maybe we can only introduce the concept of keyspace in client side and client takes care of key prefix stuff. I could be much simpler as tikv or pd does not need any change.

Reply:
We effectively do have client-side key spaces right now. But I guess some clients believe they own the entirety of the key space and this API change would encourage clients to not do that?
But this won't meet my requirements of having an automated way of ensuring there is no key conflict.

From the point of view of an abstract client-side driver these would both look the same where the user configures a namespace. And I would like the client-side to implement a sub spaces feature (spaces with a keyspace) on their own, just as is done in FoundationDB. So what you are suggesting would also be like adding only sub spaces: every client can write to the whole key space but first creates a sub space that it hopes won't conflict.

Reply:
IMO client-side key space is more flexible, like Redis on TiKV, you can define your own key spaces.

Reply:
How would it be more flexible? Ease of implementation is different than flexibility. An opposite of flexible can be brittle, and it is certainly more brittle because there is no de-coupling between logical and physical key space to prevent collisions: that must be ensured by ad-hoc mechanisms.

I do think we should encourage clients to utilize sub spaces as needed independently of this proposal.

Reply:
The proposal has been updated so that TiKV does not have knowledge of key spaces, but they are still registered in PD.

Reply:
I created a new RFC for TiKV to assist with sending prefixes that can be used for key spaces: #56. This prefix capability can be used to create client-only key spaces.
This RFC is changed now so that TiKV has no knowledge of key spaces, but they are still registered in PD so that all clients can be aware of other registered key prefixes.